File: android/src/main/java/expo/modules/facevision/FaceVisionException.kt
package expo.modules.facevision

import expo.modules.kotlin.exception.CodedException

internal object ErrorCodes {
  const val INVALID_URI = "INVALID_URI"
  const val FILE_NOT_FOUND = "FILE_NOT_FOUND"
  const val INVALID_IMAGE = "INVALID_IMAGE"
  const val DETECTION_FAILED = "DETECTION_FAILED"
  const val IMAGE_WRITE_FAILED = "IMAGE_WRITE_FAILED"
  const val INVALID_ARGUMENT = "INVALID_ARGUMENT"
  const val CAMERA_UNAVAILABLE = "CAMERA_UNAVAILABLE"
  const val CAMERA_PERMISSION_DENIED = "CAMERA_PERMISSION_DENIED"
}

/**
 * CodedException infers its code from the class name (`FooException` -> `ERR_FOO`).
 * Passing the code explicitly keeps JS on the same codes as iOS.
 */
internal class FaceVisionException(
  code: String,
  message: String,
  cause: Throwable? = null
) : CodedException(code, message, cause)

File: ios/FaceVisionException.swift
import ExpoModulesCore

internal enum FaceVisionErrorCode: String {
  case invalidUri = "INVALID_URI"
  case fileNotFound = "FILE_NOT_FOUND"
  case invalidImage = "INVALID_IMAGE"
  case detectionFailed = "DETECTION_FAILED"
  case imageWriteFailed = "IMAGE_WRITE_FAILED"
  case invalidArgument = "INVALID_ARGUMENT"
  case cameraUnavailable = "CAMERA_UNAVAILABLE"
  case cameraPermissionDenied = "CAMERA_PERMISSION_DENIED"
}

/**
 Expo's `Exception` derives its code from the class name (`FooException` -> `ERR_FOO`).
 Overriding `code` keeps JS on the stable codes documented in the TS types.
 */
internal final class FaceVisionException: Exception, @unchecked Sendable {
  private let errorCode: String
  private let message: String

  init(_ code: FaceVisionErrorCode, _ message: String, cause: Error? = nil) {
    self.errorCode = code.rawValue
    self.message = message
    super.init()
    self.cause = cause
  }

  override var code: String {
    errorCode
  }

  override var reason: String {
    message
  }
}
